parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -481,12 +495,14 @@ isEmpty0 = (== empty0)
 isEmpty :: Branch m -> Bool
 isEmpty = (== empty)
 
+-- | Perform an update over the current branch and create a new causal step.
+-- Note: this does not step history on updated child branches, see 'stepManyAt'.
 step :: Applicative m => (Branch0 m -> Branch0 m) -> Branch m -> Branch m
-step f = \case
-  Branch (Causal.One _h e) | e == empty0 -> Branch (Causal.one (f empty0))
-  b -> over history (Causal.stepDistinct f) b
+step f = runIdentity . stepM (Identity . f)
```

Comment:
I rewrote all of the pure variants in terms of their monadic counterparts using `Identity`; it felt safer to avoid the duplication wherever possible.